### DISCUSSION TITLE
Turbine type and power curve density

### DISCUSSION THREAD

**Discussion by ghost** [NatLabRockies/floris#391] [Q&A] [answered]
Hello Everyone,

I have tried to use the new version of FLORIS and I have a couple of questions, mainly related to the definition of a wind turbine.

From example 15 check_turbines, I understand that the turbine type used in the "fi.reinitialize" must be included in the 
../floris/turbine_library directory as a yaml file.

1. Is there another way to define a wind turbine? I'm thinking of a dictionary directly.

2. Is the power curve assumed to be defined at air density = 1.225? Is there a way to define the power curve for a different air density or do you plan to include it?

Thank you!

**Comment by paulf81** [comment, marked as answer]
Hi!  Thank you for these questions!

On (1), this is something we are currently thinking about.  I will come back to you an that point

On (2), if you check out the develop branch there have been some fixes which should show if you re-run example 15.  Right now FLORIS does assume the power curve is defined at 1.225, and if another air density is supplied this changes scales the effective wind speed (see line 137 in turbine.py in the develop branch).  It could be possible to replace the fixed assumption of cp/ct defined at 1.225 with an additional variable in the turbine yaml "reference density".  Would this be a valuable change?

**Reply by ghost** [reply]
Thank you very much Paul.
I think that it would be a valuable change, because if you have more defined power and thrust curve data, I imagine that they will be more reliable than through the correction of the wind speed. Honestly, I am not familiar with effective speed correction through density, nor do I know how far it is applicable.

**Comment by Bartdoekemeijer** [comment]
Additionally, for (1), note that turbine YAML files can be placed anywhere. See [this post](https://github.com/NREL/floris/issues/345#issuecomment-1062104333). You can also update turbine definitions using the `fi.reinitialize` function, like you mentioned. One example:

```
import copy
from floris.tools import FlorisInterface

# Instantiate FLORIS
fi = FlorisInterface("inputs/gch.yaml")
nrel5mw_dict = fi.floris.farm.turbine_definitions[0]  # Copy nominal turbine definition

# Now copy turbine dict but make a very tall tower
nrel5mw_tall = copy.deepcopy(nrel5mw_dict)
nrel5mw_tall["hub_height"] = 110.0

# and create a copy with a small rotor
nrel5mw_smallrotor = copy.deepcopy(nrel5mw_dict)
nrel5mw_smallrotor["rotor_diameter"] = 95.0

# Update farm: two-turbine layout, with first turbine having a tall tower, and second turbine a small rotor
fi.reinitialize(
    layout=[[0.0, 600.0], [0.0, 0.0]],
    turbine_type=[nrel5mw_tall, nrel5mw_smallrotor],
    reference_wind_height=90.0,
)
```

You can inspect the `nrel5mw_dict` dictionary yourself, but essentially it contains the following keys:
```
dict_keys(['turbine_type', 'generator_efficiency', 'hub_height', 'pP', 'pT', 'rotor_diameter', 'TSR', 'power_thrust_table'])
```
and `power_thrust_table` therein is another `dict` which contains the following keys:
```
dict_keys(['power', 'thrust', 'wind_speed'])
```
So yes, you can specify and provide dictionaries directly to `FlorisInterface` to assign turbine properties. Just make sure you cover all the required keys and format things appropriately.

**Reply by ghost** [reply]
Thank you Bart! This fully answers my question 1.

**Comment by rafmudaf** [comment]
Yet another option for the turbine inputs is to define a dictionary directly in the input file:

```yaml
farm:
  layout_x:
  - 0.0
  - 630.0
  - 1260.0
  layout_y:
  - 0.0
  - 0.0
  - 0.0
  turbine_type:
  - turbine_type: "raf_20mw"
    generator_efficiency: 1.0
    hub_height: 90.0
    pP: 1.88
    pT: 1.88
    rotor_diameter: 126.0
    TSR: 8.0
    power_thrust_table:
```

**Comment by paulf81** [comment]
hi @mapasan , I've opened pull request #398 to address this issue, does this look like what you have in mind?
